### DISCUSSION TITLE
Can I run "Replace failed" action against a removed node?

### DISCUSSION THREAD

**Discussion by philromine** [zwave-js/zwave-js-ui#3398] [Q&A] [answered]
### Checklist

- [X]  I am not using Home Assistant. Or: a developer has told me to come here.
- [X] I have checked the troubleshooting section and my problem is not described there.
- [X] I have read the changelog and my problem is not mentioned there.

### Deploy method

Manually built (git clone - yarn build - yarn install

### Z-Wave JS UI version

9.3.1

### ZwaveJS version

12.3.0

### Describe the bug

Zwave JS UI presents option to "Write the node Id and press enter if not present" when doing a Replace "Replace a failed device", but node-jwave-js packages/zwave-js/src/lib/Controller.ts calls this.nodes.getOrThrow(nodeId), rejecting the request if the node is not in the controller's this.nodes.

Ideally, the fix would be to node-jwave-js to support replace of removed nodes...

### To Reproduce

Do a Replace type inclusion, but enter a node Id of a node that is not present (has been removed).

### Expected behavior

The desired behavior (for me) is that node-zwave-js would allow a node to be replaced even if not in the controller's this.nodes.

### Additional context

I'm a retired full-stack developer, pretty skilled in TS/JS.  I switched to zwavejs from hubitat.  My install is a 15-unit motel with about 150 zwave nodes: Aeotec 700 controller, Yale locks, Zooz LR dimmers, Enbrighten motion sensors/switches, Zooz Zen51 dry-contact sensors connected to smoke detector, Aeotec home energy meters, heavy-duty switches, Alarm.com (Building 36) water valve/meters, Ecolink thermostats, etc.

I'd be glad to become a contributor if I could be of help.

**Comment by robertsLando** [comment, marked as answer]
If I remember correctly this is how zwave protocol works, it expects the node to be in controller nodes failed list in order to allow a replace so if you did an exclusion you cannot do a replace with that nodeId. So this is not an issue with zwave-js itself but I let @AlCalzone to give you a better detailed answer.

As a side note I can say that I was doing this with OpenZwave and it worked, I mean I was able to run the replace failed even against removed nodes, but zwave-js must be protocol complaint so it doesn't allow that. Maybe this could be allowed with a special flag or Driver option? @AlCalzone

**Reply by AlCalzone** [reply]
AFAIK technically it is possible by passing a non-existing node ID with the controller command, but it has to be considered a hack, and I was advised against allowing it by a former Silabs engineer.
And aside from cosmetic reasons there's IMO not much reason to do it.

**Reply by philromine** [reply]
Cool - I had locally modified Controller.ts to support it, but don't want to be hacky... I'd glanced through the code for rebuilding routes, saw some things seem to be done in node-id-order, and, as I had some no-longer-used low node id's and some good new repeaters, I guessed there might be a benefit to adding them as the low node id's... if having a low-vs-high node number makes no difference, then I understand, but if it does, I thought this might be a good way to not have to start-over, especially with a large network. I appreciate the time and thank you all for the whole zwavejs project!

**Reply by AlCalzone** [reply]
Pings at startup are no longer done in node ID order. Might make sense to do the same for rebuilding routes. But then this shouldn't be done frequently anyways, so it doesn't really matter.

**Reply by philromine** [reply]
Good to know - thanks!
